keanu-project/src/main/java/io/improbable/keanu/algorithms/mcmc/HamiltonianSampler.java

```diff
@@ -0,0 +1,280 @@
+            momentumsAtHalfTimeStep.put(vertexMomentum.getKey(), updatedMomentum);
+        }
+
+        //Set `T = T + `r.
+        for (Vertex<DoubleTensor> latent : latentVertices) {
+            final DoubleTensor nextPosition = momentumsAtHalfTimeStep.get(latent.getId()).times(halfTimeStep).plusInPlace(during.position.get(latent.getId()));
+            during.position.put(latent.getId(), nextPosition);
+            latent.setValue(nextPosition);
+        }
+
+        VertexValuePropagation.cascadeUpdate(latentVertices);
+
+        //Set `r = `r + (eps/2)dTL(`T)
+        Map<VertexId, DoubleTensor> newGradient = logProbGradient.getJointLogProbGradientWrtLatents();
+
+        for (Map.Entry<VertexId, DoubleTensor> halfTimeStepMomentum : momentumsAtHalfTimeStep.entrySet()) {
+            final DoubleTensor updatedMomentum = newGradient.get(halfTimeStepMomentum.getKey()).times(halfTimeStep).plusInPlace(halfTimeStepMomentum.getValue());
+            during.momentum.put(halfTimeStepMomentum.getKey(), updatedMomentum);
+        }
+
+        return newGradient;
+    }
+
+    private static double getLikelihoodOfLeapfrog(final double logOfMasterPAfterLeapfrog,
+                                                  final double previousLogOfMasterP,
+                                                  final Map<VertexId, DoubleTensor> leapfroggedMomentum,
+                                                  final Map<VertexId, DoubleTensor> momentumPreviousTimeStep) {
+
+        final double leapFroggedMomentumDotProduct = (0.5 * dotProduct(leapfroggedMomentum));
+        final double previousMomentumDotProduct = (0.5 * dotProduct(momentumPreviousTimeStep));
+
+        final double leapFroggedLikelihood = logOfMasterPAfterLeapfrog - leapFroggedMomentumDotProduct;
+        final double previousLikelihood = previousLogOfMasterP - previousMomentumDotProduct;
+
+        final double logLikelihoodOfLeapFrog = leapFroggedLikelihood - previousLikelihood;
+        final double likelihoodOfLeapfrog = Math.exp(logLikelihoodOfLeapFrog);
+
+        return Math.min(likelihoodOfLeapfrog, 1.0);
+    }
+
+    private static boolean shouldReject(double likelihood, KeanuRandom random) {
+        return likelihood < random.nextDouble();
+    }
+
+    private static double dotProduct(Map<VertexId, DoubleTensor> momentums) {
+        double dotProduct = 0.0;
+        for (DoubleTensor momentum : momentums.values()) {
+            dotProduct += momentum.pow(2).sum();
+        }
+        return dotProduct;
+    }
+
+    /**
+     * This is meant to be used for caching a pre-leapfrog sample. This sample
+     * will be used if the leapfrog is rejected.
+     *
+     * @param sample
+     * @param fromVertices
+     */
+    private static void putSample(Map<VertexId, ?> sample, List<? extends Vertex> fromVertices) {
+        for (Vertex<?> vertex : fromVertices) {
+            putValue(vertex, sample);
+        }
```

Comment:
You can do this instead, although you also end up having to remove some `?`s elsewhere in this file to make the compiler happy.

    private static <T> Map<VertexId, Object> putSample(List<? extends Vertex> fromVertices) {
        return fromVertices.stream().collect(Collectors.toMap(Vertex::getId, Vertex::getValue));
    }

Reply:
It's arguably preferable to having that gnarly cast in `putValue`

Reply:
Nice - having the returned map looks much clearer too